File: src/auth/auth.service.ts
import { Injectable, BadRequestException, UnauthorizedException } from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { UsersService } from '../users/users.service';
import { MailService } from '../mail/mail.service';
import { Prisma } from '@prisma/client';
import * as bcrypt from 'bcrypt';
import { createHash, randomBytes } from 'crypto';

export interface GoogleUser {
  googleId: string;
  email: string;
  fullName: string;
  avatarUrl?: string;
}

export interface JwtPayload {
  sub: string;
  email: string;
  role?: string;
}

import { RegisterDto, LoginDto } from './dto/auth.dto';

@Injectable()
export class AuthService {
  constructor(
    private readonly usersService: UsersService,
    private readonly jwtService: JwtService,
    private readonly mailService: MailService,
  ) {}

  async register(registerDto: RegisterDto) {
    const { email, password, fullName } = registerDto;
    
    // Check if user already exists
    const existingUser = await this.usersService.findByEmail(email);
    if (existingUser) {
      throw new BadRequestException('Email already in use');
    }
    
    // Hash password
    const saltRounds = 10;
    const passwordHash = await bcrypt.hash(password, saltRounds);
    
    // Create user (unverified by default)
    const user = await this.usersService.createWithPassword({
      email,
      fullName,
      passwordHash,
    });
    
    // Generate a 6-digit OTP
    const otp = Math.floor(100000 + Math.random() * 900000).toString();
    const expiry = new Date();
    expiry.setMinutes(expiry.getMinutes() + 15); // 15 minute expiry
    
    // Update user with OTP
    await this.usersService.update(user.id, {
      emailVerificationOtp: otp,
      emailVerificationOtpExpiry: expiry,
      isEmailVerified: false,
    } as Prisma.UserUpdateInput);
    
    // Send email and fail the registration flow if delivery was rejected.
    const sent = await this.mailService.sendVerificationOtp(email, fullName, otp);
    if (!sent) {
      throw new BadRequestException('Failed to send verification email. Please try again later.');
    }
    
    return {
      message: 'Registration successful. Please verify your email.',
      requiresOtp: true,
      email: user.email,
    };
  }

  async login(loginDto: LoginDto) {
    const { email, password } = loginDto;
    
    // Find user by email
    const user = await this.usersService.findByEmail(email);
    if (!user) {
      throw new UnauthorizedException('Invalid email or password');
    }
    
    // Check if user has a password (might be Google-only user)
    if (!user.passwordHash) {
      throw new UnauthorizedException('Please sign in with Google');
    }
    
    // Verify password
    const isPasswordValid = await bcrypt.compare(password, user.passwordHash);
    if (!isPasswordValid) {
      throw new UnauthorizedException('Invalid email or password');
    }
    
    // Check if email is verified
    if (!user.isEmailVerified) {
      // Generate a new OTP automatically on login attempt
      const otp = Math.floor(100000 + Math.random() * 900000).toString();
      const expiry = new Date();
      expiry.setMinutes(expiry.getMinutes() + 15);
      
      await this.usersService.update(user.id, {
        emailVerificationOtp: otp,
        emailVerificationOtpExpiry: expiry,
      } as Prisma.UserUpdateInput);
      
      await this.mailService.sendVerificationOtp(user.email, user.fullName || 'User', otp);
      
      // We throw a specific structure so the frontend knows to redirect to the OTP screen
      throw new UnauthorizedException({
        message: 'Email not verified. A new code has been sent.',
        requiresOtp: true,
        email: user.email,
      });
    }
    
    // Generate tokens
    const tokens = await this.generateTokens(user.id, user.email, user.role);
    
    return {
      user: {
        id: user.id,
        email: user.email,
        fullName: user.fullName,
        avatarUrl: user.avatarUrl,
        targetBandScore: user.targetBandScore,
        nativeLanguage: user.nativeLanguage,
        subscriptionTier: user.subscriptionTier,
        subscriptionExpiresAt: user.subscriptionExpiresAt,
        role: user.role,
      },
      ...tokens,
    };
  }

  async validateGoogleUser(googleUser: GoogleUser) {
    // Find or create user
    let user = await this.usersService.findByGoogleId(googleUser.googleId);
    let isNew = false;

    if (!user) {
      // Check if user exists with email
      user = await this.usersService.findByEmail(googleUser.email);

      if (user) {
        // Link Google account to existing user
        user = await this.usersService.linkGoogleAccount(user.id, googleUser.googleId, googleUser.avatarUrl);
      } else {
        // Create new user
        user = await this.usersService.createFromGoogle(googleUser);
        isNew = true;
      }
    }

    return { ...user, isNew };
  }
  
  async verifyOtp(email: string, otp: string) {
    const user = await this.usersService.findByEmail(email);
    
    if (!user) {
      throw new BadRequestException('User not found');
    }
    
    // Check if OTP matches and is not expired
    if (!user.emailVerificationOtp || user.emailVerificationOtp !== otp) {
      throw new BadRequestException('Invalid verification code');
    }
    
    if (user.emailVerificationOtpExpiry && new Date() > user.emailVerificationOtpExpiry) {
      throw new BadRequestException('Verification code has expired. Please request a new one.');
    }
    
    // Mark as verified and clear OTP fields
    await this.usersService.update(user.id, {
      isEmailVerified: true,
      emailVerificationOtp: null,
      emailVerificationOtpExpiry: null,
    } as Prisma.UserUpdateInput);
    
    // Return tokens to log the user in immediately
    const tokens = await this.generateTokens(user.id, user.email);
    
    return {
      message: 'Email verified successfully',
      user: {
        id: user.id,
        email: user.email,
        fullName: user.fullName,
      },
      ...tokens,
    };
  }
  
  async resendOtp(email: string) {
    const user = await this.usersService.findByEmail(email);
    
    if (!user) {
      throw new BadRequestException('If an account matches this email, a code has been sent.');
    }
    
    if (user.isEmailVerified) {
      throw new BadRequestException('Email is already verified');
    }
    
    // Generate a new 6-digit OTP
    const otp = Math.floor(100000 + Math.random() * 900000).toString();
    const expiry = new Date();
    expiry.setMinutes(expiry.getMinutes() + 15);
    
    await this.usersService.update(user.id, {
      emailVerificationOtp: otp,
      emailVerificationOtpExpiry: expiry,
    } as Prisma.UserUpdateInput);
    
    const sent = await this.mailService.sendVerificationOtp(user.email, user.fullName || 'User', otp);
    
    if (!sent) {
      throw new BadRequestException('Failed to send verification email. Please try again later.');
    }
    
    return {
      message: 'Verification code sent to your email',
    };
  }

  async forgotPassword(email: string) {
    const user = await this.usersService.findByEmail(email);

    if (!user) {
      return {
        message: 'If an account matches this email, a password reset link has been sent.',
      };
    }

    const resetToken = randomBytes(32).toString('hex');
    const resetTokenHash = this.hashToken(resetToken);
    const expiry = new Date();
    expiry.setHours(expiry.getHours() + 1);

    await this.usersService.update(user.id, {
      passwordResetTokenHash: resetTokenHash,
      passwordResetTokenExpiry: expiry,
    } as Prisma.UserUpdateInput);

    const sent = await this.mailService.sendPasswordResetEmail(
      user.email,
      user.fullName,
      resetToken,
    );

    if (!sent) {
      throw new BadRequestException('Failed to send password reset email. Please try again later.');
    }

    return {
      message: 'If an account matches this email, a password reset link has been sent.',
    };
  }

  async resetPassword(token: string, password: string) {
    const user = await this.usersService.findByPasswordResetTokenHash(this.hashToken(token));

    if (!user || !user.passwordResetTokenExpiry || user.passwordResetTokenExpiry < new Date()) {
      throw new BadRequestException('Reset link is invalid or has expired.');
    }

    const saltRounds = 10;
    const passwordHash = await bcrypt.hash(password, saltRounds);

    await this.usersService.update(user.id, {
      passwordHash,
      passwordResetTokenHash: null,
      passwordResetTokenExpiry: null,
    } as Prisma.UserUpdateInput);

    return {
      message: 'Password reset successful. You can now sign in with your new password.',
    };
  }

  async generateTokens(userId: string, email: string, role?: string) {
    const payload: JwtPayload = { sub: userId, email, role };
    
    const accessToken = this.jwtService.sign(payload);
    
    return {
      accessToken,
      expiresIn: '7d',
    };
  }

  async validateJwtPayload(payload: JwtPayload) {
    const user = await this.usersService.findById(payload.sub);
    return user;
  }

  private hashToken(token: string) {
    return createHash('sha256').update(token).digest('hex');
  }
}

File: internal/auth/service/auth_service.go
package authservice

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"math/big"
	"time"

	"github.com/Preshy-Jones/bandready-server/internal/config"
	"github.com/Preshy-Jones/bandready-server/internal/mail"
	"github.com/Preshy-Jones/bandready-server/internal/models"
	"github.com/golang-jwt/jwt/v5"
	"github.com/google/uuid"
	"golang.org/x/crypto/bcrypt"
	"gorm.io/gorm"
)

type AuthService struct {
	db   *gorm.DB
	cfg  *config.Config
	mail *mail.MailService
}

func New(db *gorm.DB, cfg *config.Config, mail *mail.MailService) *AuthService {
	return &AuthService{db: db, cfg: cfg, mail: mail}
}

type RegisterInput struct {
	Email    string
	Password string
	FullName string
}

type LoginInput struct {
	Email    string
	Password string
}

type AuthResult struct {
	Token string
	User  *models.User
	IsNew bool
}

func (s *AuthService) Register(input RegisterInput) (*AuthResult, error) {
	var existing models.User
	if err := s.db.Where("email = ?", input.Email).First(&existing).Error; err == nil {
		return nil, errors.New("email already registered")
	}

	hash, err := bcrypt.GenerateFromPassword([]byte(input.Password), 10)
	if err != nil {
		return nil, fmt.Errorf("failed to hash password: %w", err)
	}

	otp, otpExpiry := generateOTP()
	hashStr := string(hash)
	fullName := input.FullName

	user := models.User{
		ID:                         uuid.New().String(),
		Email:                      input.Email,
		PasswordHash:               &hashStr,
		FullName:                   &fullName,
		EmailVerificationOtp:       &otp,
		EmailVerificationOtpExpiry: &otpExpiry,
		Role:                       models.RoleUser,
		SubscriptionTier:           models.SubscriptionNone,
		ExamType:                   models.ExamAcademic,
	}

	if err := s.db.Create(&user).Error; err != nil {
		return nil, fmt.Errorf("failed to create user: %w", err)
	}

	go s.mail.SendVerificationOTP(user.Email, input.FullName, otp)

	token, err := s.generateToken(&user)
	if err != nil {
		return nil, err
	}

	return &AuthResult{Token: token, User: &user, IsNew: true}, nil
}

func (s *AuthService) Login(input LoginInput) (*AuthResult, error) {
	var user models.User
	if err := s.db.Where("email = ?", input.Email).First(&user).Error; err != nil {
		return nil, errors.New("invalid email or password")
	}

	if user.PasswordHash == nil {
		return nil, errors.New("please sign in with Google")
	}

	if err := bcrypt.CompareHashAndPassword([]byte(*user.PasswordHash), []byte(input.Password)); err != nil {
		return nil, errors.New("invalid email or password")
	}

	token, err := s.generateToken(&user)
	if err != nil {
		return nil, err
	}

	return &AuthResult{Token: token, User: &user}, nil
}

func (s *AuthService) VerifyOTP(email, otp string) (*AuthResult, error) {
	var user models.User
	if err := s.db.Where("email = ?", email).First(&user).Error; err != nil {
		return nil, errors.New("user not found")
	}

	if user.EmailVerificationOtp == nil || *user.EmailVerificationOtp != otp {
		return nil, errors.New("invalid OTP")
	}

	if user.EmailVerificationOtpExpiry == nil || time.Now().After(*user.EmailVerificationOtpExpiry) {
		return nil, errors.New("OTP has expired")
	}

	if err := s.db.Model(&user).Updates(map[string]interface{}{
		"is_email_verified":            true,
		"email_verification_otp":       nil,
		"email_verification_otp_expiry": nil,
	}).Error; err != nil {
		return nil, err
	}

	token, err := s.generateToken(&user)
	if err != nil {
		return nil, err
	}

	return &AuthResult{Token: token, User: &user}, nil
}

func (s *AuthService) ResendOTP(email string) error {
	var user models.User
	if err := s.db.Where("email = ?", email).First(&user).Error; err != nil {
		return errors.New("user not found")
	}

	if user.IsEmailVerified {
		return errors.New("email already verified")
	}

	otp, expiry := generateOTP()
	if err := s.db.Model(&user).Updates(map[string]interface{}{
		"email_verification_otp":        otp,
		"email_verification_otp_expiry": expiry,
	}).Error; err != nil {
		return err
	}

	fullName := ""
	if user.FullName != nil {
		fullName = *user.FullName
	}
	go s.mail.SendVerificationOTP(email, fullName, otp)
	return nil
}

func (s *AuthService) ForgotPassword(email string) error {
	var user models.User
	if err := s.db.Where("email = ?", email).First(&user).Error; err != nil {
		// Don't reveal if email exists
		return nil
	}

	token := generateResetToken()
	hash := sha256Hash(token)
	expiry := time.Now().Add(1 * time.Hour)

	if err := s.db.Model(&user).Updates(map[string]interface{}{
		"password_reset_token_hash":   hash,
		"password_reset_token_expiry": expiry,
	}).Error; err != nil {
		return err
	}

	fullName := ""
	if user.FullName != nil {
		fullName = *user.FullName
	}
	go s.mail.SendPasswordResetEmail(email, fullName, token)
	return nil
}

func (s *AuthService) ResetPassword(token, newPassword string) error {
	hash := sha256Hash(token)

	var user models.User
	if err := s.db.Where("password_reset_token_hash = ?", hash).First(&user).Error; err != nil {
		return errors.New("invalid or expired reset token")
	}

	if user.PasswordResetTokenExpiry == nil || time.Now().After(*user.PasswordResetTokenExpiry) {
		return errors.New("reset token has expired")
	}

	passwordHash, err := bcrypt.GenerateFromPassword([]byte(newPassword), 10)
	if err != nil {
		return err
	}

	hashStr := string(passwordHash)
	return s.db.Model(&user).Updates(map[string]interface{}{
		"password_hash":               hashStr,
		"password_reset_token_hash":   nil,
		"password_reset_token_expiry": nil,
	}).Error
}

func (s *AuthService) ValidateGoogleUser(googleID, email, name, avatarURL string) (*AuthResult, bool, error) {
	var user models.User
	isNew := false

	err := s.db.Where("google_id = ?", googleID).First(&user).Error
	if err == nil {
		token, err := s.generateToken(&user)
		return &AuthResult{Token: token, User: &user}, isNew, err
	}

	// Try to link to existing account by email
	err = s.db.Where("email = ?", email).First(&user).Error
	if err == nil {
		s.db.Model(&user).Updates(map[string]interface{}{
			"google_id":  googleID,
			"avatar_url": avatarURL,
		})
		token, err := s.generateToken(&user)
		return &AuthResult{Token: token, User: &user}, isNew, err
	}

	// Create new user
	isNew = true
	user = models.User{
		ID:               uuid.New().String(),
		Email:            email,
		GoogleID:         &googleID,
		FullName:         &name,
		AvatarURL:        &avatarURL,
		IsEmailVerified:  true,
		Role:             models.RoleUser,
		SubscriptionTier: models.SubscriptionNone,
		ExamType:         models.ExamAcademic,
	}

	if err := s.db.Create(&user).Error; err != nil {
		return nil, false, err
	}

	token, err := s.generateToken(&user)
	return &AuthResult{Token: token, User: &user}, isNew, err
}

func (s *AuthService) GetUserByID(id string) (*models.User, error) {
	var user models.User
	if err := s.db.First(&user, "id = ?", id).Error; err != nil {
		return nil, err
	}
	return &user, nil
}

func (s *AuthService) generateToken(user *models.User) (string, error) {
	claims := jwt.MapClaims{
		"sub":   user.ID,
		"email": user.Email,
		"role":  string(user.Role),
		"exp":   time.Now().Add(s.cfg.JWTExpiration).Unix(),
		"iat":   time.Now().Unix(),
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	return token.SignedString([]byte(s.cfg.JWTSecret))
}

func generateOTP() (string, time.Time) {
	const digits = "0123456789"
	otp := make([]byte, 6)
	for i := range otp {
		n, _ := rand.Int(rand.Reader, big.NewInt(int64(len(digits))))
		otp[i] = digits[n.Int64()]
	}
	return string(otp), time.Now().Add(15 * time.Minute)
}

func generateResetToken() string {
	b := make([]byte, 32)
	rand.Read(b)
	return hex.EncodeToString(b)
}

func sha256Hash(input string) string {
	h := sha256.Sum256([]byte(input))
	return hex.EncodeToString(h[:])
}

// otpVerificationAdded - 6-digit OTP, 15min expiry

// forgotPasswordFlow - SHA256 token, 1hr expiry
